google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java

```diff
@@ -260,6 +263,89 @@ public void downloadTo(Path path) {
     downloadTo(path, new BlobSourceOption[0]);
   }
 
+  /**
+   * Uploads the given file path to this blob using specified blob write options.
```

Comment:
I probably just don't understand how this API works, but this sounds off to me. I'd expect a static method that uploads the path to a new Blob and returns it. What does it mean to "upload to this blob"? I'm missing something here.

Reply:
The storage hierarchy is the following:
Class **BlobId**: A BlobId object includes the name of the containing bucket, the blob's name and possibly the blob's generation. If getGeneration() is null the identifier refers to the latest blob's generation.

Class **BlobInfo**: encapsulate **BlobId** and set of properties.

Class **Blob** extends **BlobInfo** and adds a number of methods to modify the blob like update, copyTo, downloadTo, etc. Objects of this class are immutable. Operations that modify the blob return a new object.

New `Blob.uploadFrom()` methods will allow those who already have a blob instance to upload some content to it as simply as `blob = blob.uploadFrom(myFile);`

> I'd expect a static method that uploads the path to a new Blob and returns it.

A static method would require an extra parameter `storage`, which is not very convenient.
As I said previously I'm going to extend the Storage interface with new methods:
```
void upload(BlobInfo blobInfo, Path path, BlobWriteOption... options);
void upload(BlobInfo blobInfo, InputStream content, BlobWriteOption... options);
public Blob create(BlobInfo blobInfo, Path path, BlobWriteOption... options);
```

Reply:
This is what's throwing me: the object is immutable but has methods to modify it? Why?

This seems to be the existing API, but it's still extremely wonky. After you call uploadFrom (or downloadTo) how does the new Blob bject returned differ from the original Blob object?

Reply:
You can consider a Blob instance like a pointer to {bucket, name, generation} and methods to operate with that pointer. Each modification of the blob causes the generation change, so existing pointer becomes outdated and the Blob instance useless(Immutability does not allow to modify the generation in place). A new object returned by uploadFrom has the new generation value pointing to the latest blob version.

Reply:
It would be nice if the docs said that, or the classes were named accurately. but c'est la vie.

https://github.com/googleapis/java-storage/issues/176

My immediate question is what exactly has changed in the object? What field has been modified and how?

Reply:
new Blob objects are constructed from data received from GCS.
In case of uploadFrom only 'generation' and 'size' will be changed. (These fields are inherited from  the BlobInfo class.)
uploadFrom  uses WriteChannel, so it has to perform a separate GET request to obtain the fresh data.
upload method performs the RPC request and obtain the data to construct new object from the response.